### PR TITLE
improve check lib, make it more accurate

### PR DESCRIPTION
**Pull request by IANTHEREAL**
### What is changed and how it works?
* split privileges checking into dump and replication privileges checking
* make version checking more accurate
* add whether to check auto-increment primary key flag in shard checking

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

### CODE REVIEW

**Comment by IANTHEREAL** [comment]
@amyangfei @csuzhangxc PTAL

**Comment by IANTHEREAL** [comment]
@july2993 PTAL

File: pkg/check/mysql_server.go [hunk 1]
```diff
@@ -33,8 +33,22 @@ func NewMySQLVersionChecker(db *sql.DB, dbinfo *dbutil.DBConfig) Checker {
 	return &MySQLVersionChecker{db: db, dbinfo: dbinfo}
 }
 
-// MinVersion is mysql minimal version required
-var MinVersion = [3]uint{5, 5, 0}
+// SupportedVersion defines the MySQL/MariaDB version that DM/syncer supports
+// * 5.6.0 <= MySQL Version < 5.8.0
+// * 10.1.2 <= Mariadb Version
+var SupportedVersion = map[string]struct {
```
Comment (july2993) [inline pkg/check/mysql_server.go:39]: s/SupportedVersion/DMSupportedVersion/ or put this in dm repo?

File: pkg/check/mysql_server.go [hunk 2]
```diff
@@ -33,8 +33,22 @@ func NewMySQLVersionChecker(db *sql.DB, dbinfo *dbutil.DBConfig) Checker {
...
+	Min MySQLVersion
+	Max MySQLVersion
+}{
+	"mysql": {
+		MySQLVersion{5, 6, 0},
+		MySQLVersion{5, 8, 0},
+	},
+	"mariadb": {
+		MySQLVersion{10, 1, 2},
+		UnlimitVersion,
```
Comment (july2993) [inline pkg/check/mysql_server.go]: suggest add a MaxVersion = {math.MaxUint8,math.MaxUint8,math.MaxUint8}
and MinVersion = {0,0,0} if need
don't specially handle UnlimitVersion in all compare func

File: pkg/check/privilege.go [hunk 3]
```diff
@@ -82,69 +117,45 @@ func (pc *SourcePrivilegeChecker) VerifyPrivileges(grants []string) *Result {
...
-		if strings.Contains(grant, "ALL PRIVILEGES") {
-			result.State = StateSuccess
-			return result
-		}
-		if strings.Contains(grant, "REPLICATION SLAVE") {
-			hasPrivilegeOfReplicationSlave = true
-		}
-		if strings.Contains(grant, "REPLICATION CLIENT") {
-			hasPrivilegeOfReplicationClient = true
+	lackOfPrivileges := make([]string, 0, 3)
```
Comment (july2993) [inline pkg/check/privilege.go]: why 3? s/3/len(expectedGrants)/ ?

File: pkg/check/mysql_server_test.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+	cases := []struct {
+		rawVersion string
+		pass       bool
+	}{
+		{"5.5.0-log", false},
+		{"5.6.0-log", true},
+		{"5.7.0-log", true},
+		{"5.8.0-log", false},
+		{"5.8.1-log", false},
```
Comment (amyangfei) [inline pkg/check/mysql_server_test.go]: what about replace `5.8.0-log` with a more real world mysql version, such as `8.0.11-log`

**Review by july2993 — COMMENTED** [review]
rest LGTM

File: pkg/check/util.go [hunk 5]
```diff
@@ -57,8 +60,12 @@ func toMySQLVersion(v string) (MySQLVersion, error) {
 	return version, nil
 }
 
-// IsAtLeast means v >= min
-func (v MySQLVersion) IsAtLeast(min MySQLVersion) bool {
+// Be means v >= min
+func (v MySQLVersion) Be(min MySQLVersion) bool {
```
Comment (amyangfei) [inline pkg/check/util.go]: Is `Be` short for `bigger than or equal to`
it seems `ge` and `gt` are used more often

File: pkg/check/privilege.go [hunk 6]
```diff
@@ -28,25 +28,23 @@ import (
...
+func NewSourceDumpPrivilegeChecker(db *sql.DB, dbinfo *dbutil.DBConfig) Checker {
+	return &SourceDumpPrivilegeChecker{db: db, dbinfo: dbinfo}
 }
 
 // Check implements the Checker interface.
-// We only check REPLICATION SLAVE, REPLICATION CLIENT, RELOAD privileges.
-// REPLICATION SLAVE and REPLICATION CLIENT are required.
-// RELOAD is strongly suggested to have.
-func (pc *SourcePrivilegeChecker) Check(ctx context.Context) *Result {
+// We only check  RELOAD, SELECT privileges.
```
Comment (amyangfei) [inline pkg/check/privilege.go]:
```suggestion
// We only check RELOAD, SELECT privileges.
```

File: pkg/check/mysql_server.go (hunk 1 above)
Comment (IANTHEREAL) [reply pkg/check/mysql_server.go:39]: if put it into dm repo, `check` would imports `dm`, `dm` would also import `check`; and it's also for syncer/checker tools. Keeping it first, I have not thought how it is not

File: pkg/check/mysql_server.go (hunk 2 above)
Comment (GregoryIan) [reply pkg/check/mysql_server.go]: it also uses `UnlimitVersion` to indicate lowest boundary

Comment (IANTHEREAL) [reply pkg/check/mysql_server.go]: ok, I just want to save a variable

File: pkg/check/util.go (hunk 5 above)
Comment (IANTHEREAL) [reply pkg/check/util.go]: I am definitely not good at writing shells.

**Comment by IANTHEREAL** [comment]
@amyangfei @july2993 PTAL again

**Review by amyangfei — COMMENTED** [review]
LGTM

**Review by july2993 — APPROVED** [review]
LGTM

**Pull request merged**